### PR TITLE
[Function] Support divide and mod functions to return NULL when divided by 0

### PR DESCRIPTION
**Pull request by zenoyang**
## Proposed changes

The original implementation returns INF when divided by 0, while non-vectorized implementation returns NULL when divided by 0. So the PR support divide and mod functions to return NULL when divided by 0.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have created an issue on (Fix #ISSUE) and described the bug/feature there in detail
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/functions/function_binary_arithmetic.h [hunk 1]
```diff
@@ -112,12 +124,14 @@ struct DecimalBinaryOperation {
 
     using ResultType = ResultType_;
     using NativeResultType = typename NativeType<ResultType>::Type;
-    using Op = std::conditional_t<
-            is_float_division,
-            DivideIntegralImpl<
-                    NativeResultType,
-                    NativeResultType>, /// substitute divide by intDiv (throw on division by zero)
-            Operation<NativeResultType, NativeResultType>>;
+//    using Op = std::conditional_t<
```
Comment (HappenLee) [inline be/src/vec/functions/function_binary_arithmetic.h:127]: 注释的代码直接删除？

File: be/src/vec/functions/divide.cpp [hunk 2]
```diff
@@ -26,15 +27,19 @@ struct DivideFloatingImpl {
     static const constexpr bool allow_decimal = true;
 
     template <typename Result = ResultType>
-    static inline Result apply(A a, B b) {
+    static inline Result apply(A a, B b, NullMap& null_map, size_t index) {
```
Comment (HappenLee) [inline be/src/vec/functions/divide.cpp:30]: 这个得根据汇编的代码看一下能不能向量化

File: be/src/vec/functions/function_binary_arithmetic_to_null_type.h [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+        bool valid = cast_both_types(
+                first_type, secord_type, [&](const auto& left, const auto& right) {
+                    using LeftDataType = std::decay_t<decltype(left)>;
+                    using RightDataType = std::decay_t<decltype(right)>;
+                    using ResultDataType =
+                            typename BinaryOperationTraits<Op, LeftDataType, RightDataType>::ResultDataType;
+                    if constexpr (!std::is_same_v<ResultDataType, InvalidType>) {
+                        if constexpr (IsDataTypeDecimal<LeftDataType> &&
+                                      IsDataTypeDecimal<RightDataType>) {
+                            constexpr bool is_multiply =
```
Comment (HappenLee) [inline be/src/vec/functions/function_binary_arithmetic_to_null_type.h:80]: 这个应该不会有乘法？

File: be/src/vec/functions/function_binary_arithmetic_to_null_type.h [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        if (!valid) {
+            LOG(FATAL) << fmt::format("Illegal types {} and {} of arguments of function {}",
+                                      arguments[0]->get_name(), arguments[1]->get_name(),
+                                      get_name());
+        }
+
+        return make_nullable(type_res);
+    }
+
+    bool use_default_implementation_for_nulls() const override { return false; }
```
Comment (HappenLee) [inline be/src/vec/functions/function_binary_arithmetic_to_null_type.h:117]: 这个应该是true吧，原有的null列可以按照默认规则处理。这样能减少你在execute里面大量处理nullable原始列的逻辑

File: be/src/vec/functions/int_div.h [hunk 5]
```diff
@@ -18,55 +18,30 @@
...
 template <typename A, typename B>
 struct DivideIntegralImpl {
     using ResultType = typename NumberTraits::ResultOfIntegerDivision<A, B>::Type;
 
     template <typename Result = ResultType>
-    static inline Result apply(A a, B b) {
-        throw_if_division_leads_to_fpe(a, b);
+    static inline Result apply(A a, B b, NullMap& null_map, size_t index) {
+        if (UNLIKELY(b == 0)) {
+            null_map[index] = 1;
```
Comment (HappenLee) [inline be/src/vec/functions/int_div.h:42]: 你这里等于0的时候不return，后面会core的

File: fe/fe-core/src/main/java/org/apache/doris/analysis/Expr.java [hunk 6]
```diff
@@ -1562,10 +1562,18 @@ public Expr findEqual(List<Expr> exprs) {
     protected Function getBuiltinFunction(
             Analyzer analyzer, String name, Type[] argTypes, Function.CompareMode mode)
             throws AnalysisException {
+        return getBuiltinFunction(analyzer, name, argTypes, mode, Function.NullableMode.DEPEND_ON_ARGUMENT);
+    }
+
+    protected Function getBuiltinFunction(Analyzer analyzer, String name, Type[] argTypes,
+                                           Function.CompareMode compareMode, Function.NullableMode nullableMode)
```
Comment (HappenLee) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/Expr.java:1569]: 这个函数啥意思？获取函数的时候还要知道是不是nullable的？这个看着有些问题啊

**Review by HappenLee — COMMENTED** [review]
有些问题，请修改一下

File: be/src/vec/functions/function_binary_arithmetic.h (hunk 1 above)
Comment (zenoyang) [reply be/src/vec/functions/function_binary_arithmetic.h:127]: ok，漏删了

File: be/src/vec/functions/function_binary_arithmetic_to_null_type.h (hunk 3 above)
Comment (zenoyang) [reply be/src/vec/functions/function_binary_arithmetic_to_null_type.h:80]: 是的，不会有乘法，我直接改成false

File: be/src/vec/functions/function_binary_arithmetic_to_null_type.h (hunk 4 above)
Comment (zenoyang) [reply be/src/vec/functions/function_binary_arithmetic_to_null_type.h:117]: ok，我改一下

File: be/src/vec/functions/int_div.h (hunk 5 above)
Comment (zenoyang) [reply be/src/vec/functions/int_div.h:42]: 会返回inf，我添加个return 0吧

File: fe/fe-core/src/main/java/org/apache/doris/analysis/Expr.java (hunk 6 above)
Comment (zenoyang) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/Expr.java:1569]: 嗯嗯，这块一开始理解错了，以为是初始化函数，我revert一下

File: be/src/vec/functions/int_div.h (hunk 5 above)
Comment (HappenLee) [reply be/src/vec/functions/int_div.h:42]: 如果返回inf的话，这个函数就很好向量化了

Comment (HappenLee) [reply be/src/vec/functions/int_div.h:42]: 你改成null_map[index] = b == 0; 然后另一个直接返回计算结果

Comment (zenoyang) [reply be/src/vec/functions/int_div.h:42]: 好的，我测试一下

File: be/src/vec/functions/function_binary_arithmetic_to_null_type.h [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+
+                        typename ColVecResult::MutablePtr col_res = nullptr;
+                        if constexpr (result_is_decimal) {
+                            ResultDataType type = decimal_result_type(left, right, is_multiply, is_division);
+                            col_res = ColVecResult::create(0, type.get_scale());
+                        } else {
+                            col_res = ColVecResult::create();
+                        }
+
+                        auto &vec_res = col_res->get_data();
```
Comment (HappenLee) [inline be/src/vec/functions/function_binary_arithmetic_to_null_type.h:189]: auto& not auto &

File: be/src/vec/functions/function_binary_arithmetic_to_null_type.h [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+                                DecimalBinaryOperation<T0, T1, Op, ResultType>,
+                                BinaryOperationImpl<T0, T1, Op<T0, T1>, ResultType>>;
+
+                        auto null_map = ColumnUInt8::create(input_rows_count, 0);
+                        auto &null_map_data = null_map->get_data();
+                        size_t argument_size = arguments.size();
+                        ColumnPtr argument_columns[argument_size];
+
+                        for (size_t i = 0; i < argument_size; ++i) {
+                            argument_columns[i] =
```
[the remark this answers is not in the document]
Comment (zenoyang) [reply be/src/vec/functions/function_binary_arithmetic_to_null_type.h:176]: 已处理

**Pull request merged**